lumicks/pylake/confocal.py

```diff
@@ -0,0 +1,241 @@
+import json
+import numpy as np
+
+from .detail.mixin import PhotonCounts
+from .detail.mixin import ExcitationLaserPower
+
+
+"""Axis label used for plotting"""
+axis_label = ("x", "y", "z")
+
+
+class BaseScan(PhotonCounts, ExcitationLaserPower):
+    """Base class for confocal scans
+
+    Parameters
+    ----------
+    name : str
+        confocal scan name
+    file : lumicks.pylake.File
+        Parent file. Contains the channel data.
+    start : int
+        Start point in the relevant info wave.
+    stop : int
+        End point in the relevant info wave.
+    json : dict
+        Dictionary containing metadata.
+    """
+
+    def __init__(self, name, file, start, stop, json):
+        self.start = start
+        self.stop = stop
+        self.name = name
+        self.json = json
+        self.file = file
+        self._cache = {}
+
+    @classmethod
+    def from_dataset(cls, h5py_dset, file):
+        """
+        Construct a confocal class from dataset.
+
+        Parameters
+        ----------
+        h5py_dset : h5py.Dataset
+            The original HDF5 dataset containing confocal scan information
+        file : lumicks.pylake.File
+            The parent file. Used to loop up channel data
+        """
+        start = h5py_dset.attrs["Start time (ns)"]
+        stop = h5py_dset.attrs["Stop time (ns)"]
+        name = h5py_dset.name.split("/")[-1]
+        try:
+            json_data = json.loads(h5py_dset[()])["value0"]
+        except KeyError:
+            # TODO => use class name
+            raise KeyError(f"Scan '{name}' is missing metadata and cannot be loaded")
+
+        return cls(name, file, start, stop, json_data)
+
+    def _get_photon_count(self, name):
+        photon_count = getattr(self.file, f"{name}_photon_count".lower())[self.start:self.stop]
+
+        try:
+            if self._has_incorrect_start(photon_count._src.start, photon_count._src.dt):
+                photon_count = getattr(self.file, f"{name}_photon_count".lower())[self.start:self.stop]
+        except AttributeError:
+            pass
+
+        return photon_count
+
+    def _plot_color(self, color, **kwargs):
+        raise NotImplementedError
+
+    def plot_red(self, **kwargs):
+        """Plot an image of the red photon channel
+
+        Parameters
+        ----------
+        **kwargs
+            Forwarded to :func:`matplotlib.pyplot.imshow`.
+        """
+        return self._plot_color("red", **kwargs) # ?? need to return here?
```

Comment:
Regarding the comment, yes, see: https://github.com/lumicks/pylake/pull/55/files

You can add that it's used to maintain a handle to images so they can be updated in existing plots.